File: sms-reply-loop-rate-limit-14107/python/twilio_reply_loop_audit.py
"""Report pairs of numbers whose traffic is an SMS reply loop.

Read only. GET requests and nothing else: give this an API Key with read access
rather than the account auth token. The repair is printed, never performed,
because this script holds a credential to an account that can send messages and
spend money.
"""
import argparse
import collections
import datetime as dt
import email.utils
import logging
import os
import sys

import requests

logging.basicConfig(level=logging.INFO, format="%(asctime)s %(levelname)s %(message)s")
log = logging.getLogger("twilio_reply_loop_audit")

HOST = "https://api.twilio.com"
BASE = HOST + "/2010-04-01"
MONITOR = "https://monitor.twilio.com/v1"

# Twilio's guard against messaging loops: 30 messages between the same two
# numbers in 30 seconds. The limit is the symptom, the loop is the bug.
LOOP_WINDOW = 30
LOOP_LIMIT = 30
ECHO_REPEATS = 4

RATE_LIMIT_ERROR = "14107"


def to_epoch(value):
    """RFC 2822 timestamp to epoch seconds, or None. Parsed rather than sliced:
    a fixed length slice of that format reads the same for every message."""
    if not value:
        return None
    try:
        return email.utils.parsedate_to_datetime(str(value)).timestamp()
    except (TypeError, ValueError):
        return None


def densest_window(stamps, window=LOOP_WINDOW):
    """Largest number of timestamps falling inside any `window` second span.

    Pure. A sliding window rather than clock buckets, because a loop starting at
    12:00:45 splits evenly across two minute buckets and disappears.
    """
    xs = sorted(s for s in stamps if s is not None)
    best = start = 0
    for i, t in enumerate(xs):
        while t - xs[start] > window:
            start += 1
        best = max(best, i - start + 1)
    return best


def classify_pair(messages, window=LOOP_WINDOW, limit=LOOP_LIMIT,
                  echo_repeats=ECHO_REPEATS):
    """Classify the traffic between one pair of numbers.

    `messages` is both directions merged: dicts with `direction`, `body` and
    `at`, the last being epoch seconds. Pure, so the density rule and the
    direction rule can be tested without a network. Returns (state, detail).
    """
    messages = messages or []
    if not messages:
        return ("quiet", "no messages between this pair in the window.")

    peak = densest_window([m.get("at") for m in messages], window)
    directions = [str(m.get("direction") or "") for m in messages]
    inbound = any(d.startswith("inbound") for d in directions)
    outbound = any(d.startswith("outbound") for d in directions)
    auto = any(d == "outbound-reply" for d in directions)
    bodies = collections.Counter(str(m.get("body") or "").strip()
                                 for m in messages if str(m.get("body") or "").strip())
    repeats = max(bodies.values()) if bodies else 0

    handwriting = (" Some of these are direction outbound-reply, which means "
                   "TwiML generated them in answer to an inbound message: that "
                   "is the loop's own handwriting." if auto else "")

    if peak >= limit and inbound and outbound:
        return ("reply-loop",
                "%d messages inside %d seconds, in both directions, with one "
                "body repeated %d times. That is the ceiling 14107 enforces, "
                "and the repair is in the inbound handler.%s"
                % (peak, window, repeats, handwriting))

    if peak >= limit:
        return ("one-way-burst",
                "%d messages inside %d seconds and all in one direction: a send "
                "loop or a retry storm in your own code, not a reply loop. Same "
                "error code, different repair." % (peak, window))

    if inbound and outbound and repeats >= echo_repeats:
        return ("echo",
                "one body repeated %d times in both directions, peaking at %d "
                "messages inside %d seconds. Under the limit, so nothing has "
                "failed and nothing will stop it either.%s"
                % (repeats, peak, window, handwriting))

    return ("normal",
            "%d message(s), peaking at %d inside %d seconds." % (len(messages), peak, window))


def get(session, url, **params):
    r = session.get(url, params=params, timeout=30)
    if r.status_code in (401, 403):
        raise SystemExit("%d from Twilio: check TWILIO_ACCOUNT_SID and that the "
                         "API key belongs to that account with read access"
                         % r.status_code)
    r.raise_for_status()
    return r.json()


def rate_limit_alerts(session, days, max_alerts):
    """Alerts carrying 14107. There is no error code filter on the request, so
    the sweep is filtered here; alerts are retained 30 days."""
    since = (dt.date.today() - dt.timedelta(days=days)).isoformat()
    url = "%s/Alerts" % MONITOR
    params = {"LogLevel": "error", "StartDate": since, "PageSize": 100}
    out = []
    while url and len(out) < max_alerts:
        page = get(session, url, **params)
        for alert in page.get("alerts", []):
            if str(alert.get("error_code") or "") == RATE_LIMIT_ERROR:
                out.append(alert)
        url = (page.get("meta") or {}).get("next_page_url")
        params = {}
    return out[:max_alerts]


def pair_from_alert(session, account, alert):
    """The two numbers behind one alert, via its Message resource.

    The list response does not carry request variables; only the single-alert
    fetch does. Resolving through the message SID keeps this to one small GET.
    """
    sid = str(alert.get("resource_sid") or "")
    if not sid.startswith(("SM", "MM")):
        return None
    msg = get(session, "%s/Accounts/%s/Messages/%s.json" % (BASE, account, sid))
    a, b = msg.get("from"), msg.get("to")
    return (a, b) if a and b else None


def conversation(session, account, a, b, days, max_messages):
    """Both halves of a conversation, merged and sorted.

    The Messages list filters To and From independently and cannot express
    "between these two numbers", so this is two queries. Half a conversation is
    exactly the half that makes a loop look like a flood.
    """
    since = (dt.date.today() - dt.timedelta(days=days)).isoformat()
    rows = []
    for sender, recipient in ((a, b), (b, a)):
        url = "%s/Accounts/%s/Messages.json" % (BASE, account)
        params = {"From": sender, "To": recipient, "DateSent>": since, "PageSize": 1000}
        while url and len(rows) < max_messages:
            page = get(session, url, **params)
            for m in page.get("messages", []):
                rows.append({"direction": m.get("direction"), "body": m.get("body"),
                             "at": to_epoch(m.get("date_created") or m.get("date_sent"))})
            nxt = page.get("next_page_uri")
            url, params = (HOST + nxt) if nxt else None, {}
    rows.sort(key=lambda m: m["at"] or 0)
    return rows


def main():
    ap = argparse.ArgumentParser(description=__doc__)
    ap.add_argument("--days", type=int, default=3,
                    help="window in days; alerts are retained 30")
    ap.add_argument("--max-alerts", type=int, default=200,
                    help="stop after this many 14107 alerts")
    ap.add_argument("--max-messages", type=int, default=4000,
                    help="stop paging a pair's history after this many rows")
    args = ap.parse_args()

    account = os.environ.get("TWILIO_ACCOUNT_SID")
    key = os.environ.get("TWILIO_API_KEY")
    secret = os.environ.get("TWILIO_API_SECRET")
    if not (account and key and secret):
        log.error("set TWILIO_ACCOUNT_SID, TWILIO_API_KEY and TWILIO_API_SECRET "
                  "(an API Key with read access, not the auth token)")
        return 2

    session = requests.Session()
    session.auth = (key, secret)

    alerts = rate_limit_alerts(session, args.days, args.max_alerts)
    if not alerts:
        log.info("no %s alerts in the last %d day(s)", RATE_LIMIT_ERROR, args.days)
        return 0

    pairs = set()
    for alert in alerts:
        pair = pair_from_alert(session, account, alert)
        if pair:
            pairs.add(tuple(sorted(pair)))
        else:
            log.warning("alert %s does not point at a message; fetch %s/Alerts/%s "
                        "for its request variables", alert.get("sid"), MONITOR,
                        alert.get("sid"))

    bad = 0
    for a, b in sorted(pairs):
        rows = conversation(session, account, a, b, args.days, args.max_messages)
        state, detail = classify_pair(rows)
        line = "%-14s %s <-> %s  %s" % (state, a, b, detail)
        if state in ("normal", "quiet"):
            log.info(line)
            continue
        bad += 1
        log.warning(line)
        log.warning("  repair: dedupe on peer plus body inside a short window in "
                    "the inbound handler, refuse to reply to your own numbers, "
                    "and audit every <Message> action URL and <Redirect> target "
                    "for cycles. Raising the rate limit buys a longer loop.")

    log.info("%d pair(s) examined from %d alert(s), %d looping",
             len(pairs), len(alerts), bad)
    return 1 if bad else 0


if __name__ == "__main__":
    sys.exit(main())

File: sms-reply-loop-rate-limit-14107/node/twilio-reply-loop-audit.mjs
/**
 * Report pairs of numbers whose traffic is an SMS reply loop.
 *
 * Read only. GET requests and nothing else: give this an API Key with read
 * access rather than the account auth token. The repair is printed, never
 * performed.
 */
const HOST = 'https://api.twilio.com';
const BASE = `${HOST}/2010-04-01`;
const MONITOR = 'https://monitor.twilio.com/v1';

// Twilio's guard against messaging loops: 30 messages between the same two
// numbers in 30 seconds. The limit is the symptom, the loop is the bug.
const LOOP_WINDOW = 30;
const LOOP_LIMIT = 30;
const ECHO_REPEATS = 4;

const RATE_LIMIT_ERROR = '14107';

/** RFC 2822 timestamp to epoch seconds, or null. */
export function toEpoch(value) {
  const ms = Date.parse(value ?? '');
  return Number.isNaN(ms) ? null : ms / 1000;
}

/**
 * Largest number of timestamps falling inside any `window` second span. Pure. A
 * sliding window rather than clock buckets, because a loop starting at 12:00:45
 * splits evenly across two minute buckets and disappears.
 */
export function densestWindow(stamps, window = LOOP_WINDOW) {
  const xs = stamps.filter((s) => s !== null && s !== undefined).sort((p, q) => p - q);
  let best = 0;
  let start = 0;
  for (let i = 0; i < xs.length; i += 1) {
    while (xs[i] - xs[start] > window) start += 1;
    best = Math.max(best, i - start + 1);
  }
  return best;
}

/**
 * Classify the traffic between one pair of numbers. `messages` is both
 * directions merged, each with direction, body and `at` in epoch seconds. Pure.
 * Returns [state, detail].
 */
export function classifyPair(messages, window = LOOP_WINDOW, limit = LOOP_LIMIT,
                             echoRepeats = ECHO_REPEATS) {
  const rows = messages ?? [];
  if (!rows.length) return ['quiet', 'no messages between this pair in the window.'];

  const peak = densestWindow(rows.map((m) => m.at), window);
  const directions = rows.map((m) => String(m.direction ?? ''));
  const inbound = directions.some((d) => d.startsWith('inbound'));
  const outbound = directions.some((d) => d.startsWith('outbound'));
  const auto = directions.some((d) => d === 'outbound-reply');

  const bodies = new Map();
  for (const m of rows) {
    const body = String(m.body ?? '').trim();
    if (body) bodies.set(body, (bodies.get(body) ?? 0) + 1);
  }
  const repeats = bodies.size ? Math.max(...bodies.values()) : 0;

  const handwriting = auto
    ? ' Some of these are direction outbound-reply, which means TwiML generated ' +
      "them in answer to an inbound message: that is the loop's own handwriting."
    : '';

  if (peak >= limit && inbound && outbound) {
    return ['reply-loop',
      `${peak} messages inside ${window} seconds, in both directions, with one ` +
      `body repeated ${repeats} times. That is the ceiling 14107 enforces, and ` +
      `the repair is in the inbound handler.${handwriting}`];
  }

  if (peak >= limit) {
    return ['one-way-burst',
      `${peak} messages inside ${window} seconds and all in one direction: a ` +
      'send loop or a retry storm in your own code, not a reply loop. Same ' +
      'error code, different repair.'];
  }

  if (inbound && outbound && repeats >= echoRepeats) {
    return ['echo',
      `one body repeated ${repeats} times in both directions, peaking at ${peak} ` +
      `messages inside ${window} seconds. Under the limit, so nothing has ` +
      `failed and nothing will stop it either.${handwriting}`];
  }

  return ['normal',
    `${rows.length} message(s), peaking at ${peak} inside ${window} seconds.`];
}

function authHeader(key, secret) {
  return `Basic ${Buffer.from(`${key}:${secret}`).toString('base64')}`;
}

async function get(auth, url, params = {}) {
  const u = new URL(url);
  for (const [k, v] of Object.entries(params)) u.searchParams.set(k, v);
  const res = await fetch(u, { headers: { Authorization: auth } });
  if (res.status === 401 || res.status === 403) {
    throw new Error(`${res.status} from Twilio: check TWILIO_ACCOUNT_SID and ` +
                    'that the API key belongs to that account with read access');
  }
  if (!res.ok) throw new Error(`${res.status} from ${u.pathname}`);
  return res.json();
}

async function rateLimitAlerts(auth, days, maxAlerts) {
  const since = new Date(Date.now() - days * 86400000).toISOString().slice(0, 10);
  let url = `${MONITOR}/Alerts`;
  let params = { LogLevel: 'error', StartDate: since, PageSize: 100 };
  const out = [];
  while (url && out.length < maxAlerts) {
    const page = await get(auth, url, params);
    for (const alert of page.alerts ?? []) {
      if (String(alert.error_code ?? '') === RATE_LIMIT_ERROR) out.push(alert);
    }
    url = page.meta?.next_page_url ?? null;
    params = {};
  }
  return out.slice(0, maxAlerts);
}

async function pairFromAlert(auth, account, alert) {
  const sid = String(alert.resource_sid ?? '');
  if (!sid.startsWith('SM') && !sid.startsWith('MM')) return null;
  const msg = await get(auth, `${BASE}/Accounts/${account}/Messages/${sid}.json`);
  return msg.from && msg.to ? [msg.from, msg.to] : null;
}

/** Both halves of a conversation. The list filters To and From independently. */
async function conversation(auth, account, a, b, days, maxMessages) {
  const since = new Date(Date.now() - days * 86400000).toISOString().slice(0, 10);
  const rows = [];
  for (const [sender, recipient] of [[a, b], [b, a]]) {
    let url = `${BASE}/Accounts/${account}/Messages.json`;
    let params = { From: sender, To: recipient, 'DateSent>': since, PageSize: 1000 };
    while (url && rows.length < maxMessages) {
      const page = await get(auth, url, params);
      for (const m of page.messages ?? []) {
        rows.push({ direction: m.direction, body: m.body,
                    at: toEpoch(m.date_created ?? m.date_sent) });
      }
      url = page.next_page_uri ? HOST + page.next_page_uri : null;
      params = {};
    }
  }
  rows.sort((p, q) => (p.at ?? 0) - (q.at ?? 0));
  return rows;
}

function flagValue(name, fallback) {
  const i = process.argv.indexOf(name);
  return i === -1 ? fallback : Number(process.argv[i + 1]);
}

async function main() {
  const account = (process.env.TWILIO_ACCOUNT_SID || "dummy-twilio-account-sid");
  const key = (process.env.TWILIO_API_KEY || "dummy-twilio-api-key");
  const secret = (process.env.TWILIO_API_SECRET || "dummy-twilio-api-secret");
  if (!account || !key || !secret) {
    console.error('set TWILIO_ACCOUNT_SID, TWILIO_API_KEY and TWILIO_API_SECRET ' +
                  '(an API Key with read access, not the auth token)');
    process.exitCode = 2;
    return;
  }
  const auth = authHeader(key, secret);
  const days = flagValue('--days', 3);

  const alerts = await rateLimitAlerts(auth, days, flagValue('--max-alerts', 200));
  if (!alerts.length) {
    console.log(`no ${RATE_LIMIT_ERROR} alerts in the last ${days} day(s)`);
    return;
  }

  const pairs = new Set();
  for (const alert of alerts) {
    const pair = await pairFromAlert(auth, account, alert);
    if (pair) pairs.add([...pair].sort().join('|'));
    else {
      console.warn(`alert ${alert.sid} does not point at a message; fetch ` +
                   `${MONITOR}/Alerts/${alert.sid} for its request variables`);
    }
  }

  let bad = 0;
  for (const joined of [...pairs].sort()) {
    const [a, b] = joined.split('|');
    const rows = await conversation(auth, account, a, b, days,
                                    flagValue('--max-messages', 4000));
    const [state, detail] = classifyPair(rows);
    const line = `${state.padEnd(14)} ${a} <-> ${b}  ${detail}`;
    if (state === 'normal' || state === 'quiet') { console.log(line); continue; }
    bad += 1;
    console.warn(line);
    console.warn('  repair: dedupe on peer plus body inside a short window in the ' +
                 'inbound handler, refuse to reply to your own numbers, and audit ' +
                 'every <Message> action URL and <Redirect> target for cycles. ' +
                 'Raising the rate limit buys a longer loop.');
  }

  console.log(`${pairs.size} pair(s) examined from ${alerts.length} alert(s), ${bad} looping`);
  process.exitCode = bad ? 1 : 0;
}

// Only run when invoked directly. The test file imports this module, and without
// the guard main() would run there too, fail on the missing credentials and set a
// non-zero exit code that fails the whole test file even as every test passes.
if (import.meta.url === `file://${process.argv[1]}`) {
  main().catch((err) => { console.error(err.message); process.exitCode = 2; });
}
